blob/gcsblob: BeforeWrite asFunc call order has affect

# Describe the bug

The order which `asFunc` is called to get `*storage.Writer` and `**storage.ObjectHandle` seems to have affect.

### To Reproduce

```go
package main

import (
	"context"
	"log"

	"cloud.google.com/go/storage"
	"gocloud.dev/blob"
	_ "gocloud.dev/blob/gcsblob"
)

func main() {
	// change this
	bucketURL := "gs://test"
	blobKey := "test"
	encryptionKey := make([]byte, 32)

	bucket, err := blob.OpenBucket(context.Background(), bucketURL)
	if err != nil {
		panic(err)
	}

	var writerOpts *blob.WriterOptions
	writerOpts = &blob.WriterOptions{
		BeforeWrite: func(asFunc func(interface{}) bool) error {
			var gcsObjectHandler **storage.ObjectHandle
			var gcsWriter *storage.Writer

			asFunc(&gcsWriter)

			if asFunc(&gcsObjectHandler) {
				*gcsObjectHandler = (**gcsObjectHandler).Key(encryptionKey)
			}

			// If called after instead it works
			//asFunc(&gcsWriter)

			return nil
		},
	}

	blobWriter, err := bucket.NewWriter(context.Background(), blobKey, writerOpts)
	if err != nil {
		panic(err)
	}
	blobWriter.Close()

	readerOpts := &blob.ReaderOptions{
		BeforeRead: func(asFunc func(interface{}) bool) error {
			var o **storage.ObjectHandle
			if asFunc(&o) {
				*o = (**o).Key(encryptionKey)
			}

			return nil
		},
	}
	blobReader, err := bucket.NewReader(context.Background(), blobKey, readerOpts)
	if err != nil {
		panic(err)
	}
	blobReader.Close()
}

```

```sh
r$ go run gcsblob_beforewrite_asfunc_order.go
panic: blob (key "test") (code=Unknown): googleapi: got HTTP response code 400 with body: <?xml version='1.0' encoding='UTF-8'?><Error><Code>ResourceNotEncryptedWithCustomerEncryptionKey</Code><Message>The resource is not encrypted with a customer encryption key.</Message><Details>The requested object is not encrypted by a customer-supplied encryption key.</Details></Error>

goroutine 1 [running]:
main.main()
	/Users/wader/src/gosal/repro/gcsblob_beforewrite_asfunc_order.go:57 +0x2d7
exit status 2
```

## Expected behavior

That the key is used when writing the blob.

### Version

v0.19.0

### Additional context

Now when poking around in the gcsblob code a bit i noticed that there is an explicit check if the writer has been created or not before returning the ObjectHandle at https://github.com/google/go-cloud/blob/master/blob/gcsblob/gcsblob.go#L516

So i guess there is a good reason for this? is it documented somewhere?

I noticed this when implementing a service that is suppose to be compatible with another service that write gcs blobs with case sensitive metadata keys (need writer to mess with the metadata keys) and uses encryption keys.

--- a/blob/gcsblob/gcsblob.go
+++ b/blob/gcsblob/gcsblob.go
@@ -41,10 +41,10 @@
 //  - ListObject: storage.ObjectAttrs
 //  - ListOptions.BeforeList: *storage.Query
 //  - Reader: *storage.Reader
-//  - ReaderOptions.BeforeRead: **storage.ObjectHandle, *storage.Reader
+//  - ReaderOptions.BeforeRead: **storage.ObjectHandle, *storage.Reader (if accessing both, must be in that order)
 //  - Attributes: storage.ObjectAttrs
-//  - CopyOptions.BeforeCopy: *CopyObjectHandles, *storage.Copier
-//  - WriterOptions.BeforeWrite: **storage.ObjectHandle, *storage.Writer
+//  - CopyOptions.BeforeCopy: *CopyObjectHandles, *storage.Copier (if accessing both, must be in that order)
+//  - WriterOptions.BeforeWrite: **storage.ObjectHandle, *storage.Writer (if accessing both, must be in that order)
 package gcsblob // import "gocloud.dev/blob/gcsblob"
 
 import (

